src/ui/connect/connector.ts

```diff
@@ -1,107 +1,137 @@
 
 export default class Connector implements ExtensionActions {
-    private port: chrome.runtime.Port;
-    private counter: number;
+    private changeSubscribers: Set<(data: ExtensionData) => void>;
 
     constructor() {
-        this.counter = 0;
-        this.port = chrome.runtime.connect({name: 'ui'});
+        this.changeSubscribers = new Set();
     }
 
-    private getRequestId() {
-        return ++this.counter;
+    private async sendRequest<T>(request: Message) {
+        return new Promise<T>((resolve, reject) => {
+            chrome.runtime.sendMessage({from: 'ui', ...request}, ({data, error}) => {
+                if (error) {
+                    reject(error);
+                } else {
+                    resolve(data);
+                }
+            });
+        });
     }
 
-    private async sendRequest<T>(request: Message, executor: (response: Message, resolve: (data?: T) => void, reject: (error: Error) => void) => void) {
-        const id = this.getRequestId();
+    private async firefoxSendRequestWithResponse<T>(name: string) {
         return new Promise<T>((resolve, reject) => {
-            const listener = ({id: responseId, ...response}: Message) => {
-                if (responseId === id) {
-                    executor(response, resolve, reject);
-                    this.port.onMessage.removeListener(listener);
+            const dataPort = chrome.runtime.connect({name});
+            dataPort.onDisconnect.addListener(() => reject());
+            dataPort.onMessage.addListener(({data, error}) => {
+                if (error) {
+                    reject(error);
+                } else {
+                    resolve(data);
                 }
-            };
-            this.port.onMessage.addListener(listener);
-            this.port.postMessage({...request, id});
+                dataPort.disconnect();
+            });
         });
     }
 
     async getData() {
-        return await this.sendRequest<ExtensionData>({type: 'get-data'}, ({data}, resolve) => resolve(data));
+        if (isFirefox) {
+            return await this.firefoxSendRequestWithResponse<ExtensionData>('ui-get-data');
+        }
+        return await this.sendRequest<ExtensionData>({type: 'get-data'});
     }
 
     async getActiveTabInfo() {
-        return await this.sendRequest<TabInfo>({type: 'get-active-tab-info'}, ({data}, resolve) => resolve(data));
+        if (isFirefox) {
+            return await this.firefoxSendRequestWithResponse<TabInfo>('ui-get-active-tab-info');
+        }
+        return await this.sendRequest<TabInfo>({type: 'get-active-tab-info'});
+    }
+
+    private onChangesReceived = ({from, type, data}: Message) => {
+        if (from === 'background' && type === 'changes') {
+            this.changeSubscribers.forEach((callback) => callback(data));
+        }
+    };
+
+    private sendMessage(data?: any) {
```

Comment:
I think `?` should be skipped here, it sends something anyway.

Reply:
Done!